### ISSUE TITLE
False negative PossiblyNullArrayOffset using array_key_first/last

### ISSUE THREAD

**Issue by ygottschalk** [vimeo/psalm#3810]
https://psalm.dev/r/297ce008e2

caused by @ignore-nullable at
https://github.com/vimeo/psalm/blame/master/src/Psalm/Internal/Stubs/CoreGenericFunctions.phpstub#L152 and
https://github.com/vimeo/psalm/blame/master/src/Psalm/Internal/Stubs/CoreGenericFunctions.phpstub#L165

Btw: there isn't any method to give a more 'narrow' return type binding depending on a more 'narrow' param type, is it?
Something like:
```
/**
 * @param array<TKey, mixed> $arr
 *
 * @return TKey|null
 * @narrow-return null if type($arr) ~ empty-array<TKey, mixed>
 * @narrow-return !null if type($arr) ~ non-empty-array<TKey, mixed>
 */
function array_key_first($arr){}
```
Basically `@assert-if` but 'reversed'... instead of narrowing the type of a parameter based on the return of the function, narrowing the type of the return value of the function based on the parameter type

**Comment by weirdan** [comment]
> Basically @assert-if but 'reversed'... instead of narrowing the type of a parameter based on the return of the function, narrowing the type of the return value of the function based on the parameter type

Sounds like conditional return type to me, see https://psalm.dev/docs/annotating_code/type_syntax/conditional_types/

**Comment by weirdan** [comment]
Yeah, you can achieve this with conditional types: https://psalm.dev/r/624f1f06ee

**Comment by muglug** [comment]
mind PRing?

**Comment by ygottschalk** [comment]
will do
